spec/index.md

```diff
@@ -598,8 +599,129 @@ MISSING
+memory slots increases gas cost.
+
+Each data type described in section 5 can be stored in these locations. How data is stored is 
+described in this section.
+
+### 6.1. Stack
+
+The following can be stored on the stack:
+
+- base type values
+- pointers to sequence type values
+
+The size of values stored on the stack must not exceed 256 bits. Since all base types are less than
+or equal to 256 bits in size, we store them on the stack. Pointers to values stored in memory may 
+also be stored on the stack.
+
+Example:
+
+```python
+# function scope
+foo: u256 = 42 # foo is stored on the stack
+bar: u256[100] # bar is a memory pointer stored on the stack
+```
+
+### 6.2. Storage
+
+All data types can be stored in storage.
+
+#### 6.2.1. Constant size values in storage
+
+Storage pointers for constant size values are determined at compile time.  
+
+Example:
+
+```python
+# contract scope
+foo: u256 # foo is assigned a static pointer by the compiler
+```
+
+The value of a base type in storage is found by simply loading the value from storage at the
+given pointer.
+
+To find an element inside of a sequence type, the relative location of the element is added to the
+given pointer.
+
+#### 6.2.2. Maps in storage
+
+Maps are not assigned pointers, because they do not have a location in storage. They are instead 
+assigned a nonce that is used to derive the location of keyed values during runtime.
+
+Example:
+
+``` python
+# contract scope
+bar: map<address, u256> # bar is assigned a static index by the compiler
+baz: map<address, map<address, u256>> # baz is assigned a static index by the compiler
+```
+
+The expression `bar[0x00]` would resolve to the hash of both bar's nonce and the key value 
+.i.e. `keccak256(<bar nonce>, 0x00)`. Similarly, the expression `baz[0x00][0x01]` would resolve to 
+a nested hash i.e. `keccak256(keccak256(<baz nonce>, 0x00), 0x01)`.
+
+### 6.3. Memory
+
+Only sequence types can be stored in memory.
+
+The first memory slot (`0x00`) is used to keep track of the lowest available memory slot. Newly 
+allocated segments begin at the value given by this slot. When more memory has been allocated,
+this value stored in `0x00` in increased.
+
+#### 6.3.1. Sequence types in memory
+
+Sequence type values may exceed the 256-bit stack slot size, so we store them in memory and 
+reference them using pointers kept on the stack.
+
+Example:
+
+```python
+# function scope
+foo: u256[100] # foo is a pointer that references 100 * 32 bytes in memory. 
+```
+
+To find an element inside of a sequence type, the relative location of the element is added to the
+given pointer.
+
+### 6.4. Function calls
+
+Constant size values stored on the stack or in memory can be passed into and returned by functions.
+
+Memory is owned by a function if it is allocated in the function call or in the function's body.
+When a function exits, all memory owned by the function is freed.
```

Comment:
I like that we free the memory on function exit. I noticed that Solidity [does not do that](https://solidity.readthedocs.io/en/v0.7.4/internals/layout_in_memory.html):

>memory is never freed (this might change in the future).

I'd love to talk about this some more in our next weekly call.

Reply:
I removed the stuff about memory ownership and de-allocation, since gas-savings are questionable and it complicates compiler implementation.